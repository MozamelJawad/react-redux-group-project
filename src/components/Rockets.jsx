import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchRockets, rocketsReserved } from '../redux/rockets/rocketsSlice';
import '../redux/rockets/rockets.css';

function Rockets() {
  const { rockets, isLoading, error } = useSelector((store) => store.rockets);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchRockets());
  }, [dispatch]);

  if (isLoading) {
    return <div className="loading">Loading ... </div>;
  }

  if (error) {
    return (
      <div className="error">
        Error:
        { error }
      </div>
    );
  }

  const handleClick = (id) => {
    dispatch(rocketsReserved(id));
  };

  return (
    <main>
      <ul className="rocketsList">
        {
      rockets.map((rocket) => {
        const {
          id, name, description, image, reserved,
        } = rocket;
        return (
          <li key={id}>
            <img className="rocketImg" src={image} alt="" />
            <div className="rocketDetails">
              <h2 className="rocketName">{name}</h2>
              <p className="rocketDesc">
                {reserved && <button type="button" className="reseved-btn"> Reserved</button>}
                {' '}
&nbsp;
                { description }
              </p>

              <button
                type="button"
                className={`reserveBtn ${reserved ? 'booked' : 'noBooked'}`}
                onClick={() => handleClick({ id })}
              >
                { reserved ? 'Cancel Reservation' : 'Reserve Rockets'}
              </button>
            </div>
          </li>
        );
      })
      }
      </ul>
    </main>
  );
}

export default Rockets;
