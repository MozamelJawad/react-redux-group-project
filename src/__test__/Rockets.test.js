import { configureStore } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import rocketsReducer from '../redux/rockets/rocketsSlice';
import Rockets from '../components/Rockets';

describe('Render Rockets view correctly', () => {
  test('Render rockets componect with actions correctly', () => {
    const store = configureStore({
      reducer: {
        rockets: rocketsReducer,
      },
      initialState: {
        rockets: {
          rockets: [
            {
              id: 1, name: 'Rocket A', image: 'image1.jpg', description: 'Rocket A Description', reserved: true,
            },
            {
              id: 2, name: 'Rocket B', image: 'image2.jpg', description: 'Rocket B Description', reserved: false,
            },
          ],
          isLoading: false,
          error: '',
        },
      },
    });

    const { container } = render(
      <Provider store={store}>
        <Router>
          <Rockets />
        </Router>
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});
