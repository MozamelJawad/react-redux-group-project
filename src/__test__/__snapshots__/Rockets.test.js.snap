// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Rockets view correctly Render rockets componect with actions correctly 1`] = `
<div>
  <main>
    <ul
      class="rocketsList"
    />
  </main>
</div>
`;

exports[`Render Rockets view correctly Render rockets correctly 1`] = `
<div>
  <main>
    <ul
      class="rocketsList"
    />
  </main>
</div>
`;

exports[`Render profile components correctly Render joined mission and reserved rockets 1`] = `
<div>
  <main>
    <ul
      class="rocketsList"
    />
  </main>
</div>
`;
